xarray/core/combine.py

```diff
@@ -456,107 +146,372 @@ def _combine_nd(combined_ids, concat_dims, data_vars='all',
         keys, which specify position within the desired final combined result.
     concat_dims : sequence of str
         The dimensions along which the datasets should be concatenated. Must be
-        in order, and the length must match
+        in order, and the length must match the length of the tuples used as
+        keys in combined_ids. If the string is a dimension name then concat
+        along that dimension, if it is None then merge.
 
     Returns
     -------
     combined_ds : xarray.Dataset
     """
 
-    # Perform N-D dimensional concatenation
+    tile_id, ds = list(combined_ids.items())[0]
+
+    n_dims = len(tile_id)
+    if len(concat_dims) != n_dims:
+        raise ValueError("concat_dims has length {} but the datasets "
+                         "passed are nested in a {}-dimensional structure"
+                         .format(str(len(concat_dims)), str(n_dims)))
```

Comment:
You can omit the extra calls to `str()`. `.format()` already does that by default when you write `{}`, so this is exactly equivalent: 
```suggestion
                         .format(len(concat_dims), n_dims))
```